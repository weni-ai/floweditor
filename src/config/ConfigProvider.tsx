import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR =
  'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<
  ConfigProviderProps
> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType,
  };

  constructor(props: ConfigProviderProps) {
    super(props);

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config,
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
