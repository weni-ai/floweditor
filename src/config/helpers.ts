import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  Types,
  VISIBILITY_ONLINE,
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';
import i18n from './i18n';
import { NodeEditorSettings } from '../store/nodeEditor';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (
  operators: Operator[],
  config: FlowEditorConfig,
): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (
  typeConfigs: Type[],
  config: FlowEditorConfig,
  nodeSettings: NodeEditorSettings,
): Type[] => {
  return filterVisibility(typeConfigs, config, nodeSettings).map(
    (typeConfig: Type) => {
      if (typeConfig.new) {
        return {
          ...typeConfig,
          description: `${typeConfig.description} (${i18n.t('new', 'New')})`,
        };
      }

      return typeConfig;
    },
  );
};

const filterVisibility = (
  items: FlowTypeVisibility[],
  config: FlowEditorConfig,
  nodeSettings?: NodeEditorSettings,
): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    if (nodeSettings) {
      const typeItem = item as Type;
      if (typeItem.type === Types.call_brain) {
        // check if we can show the call brain action
        // if we are not a ghost node
        // and if we are in a multi action node, we can't show the call brain action
        if (
          nodeSettings.originalNode.node.actions.length >= 1 &&
          !nodeSettings.originalNode.ghost
        ) {
          console.log('filter', nodeSettings);
          return false;
        }
      }
    }

    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (
        !(config.filters || []).find((name: string) => name === item.filter)
      ) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return (
      item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1
    );
  });
};
