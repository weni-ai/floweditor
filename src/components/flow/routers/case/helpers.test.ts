import {
  getCategoryName,
  getMinMax,
  isFloat,
  isInt,
  parseNum,
  prefix,
  strContainsNum,
  validateCase,
} from 'components/flow/routers/case/helpers';
import { Operators } from 'config/interfaces';
import { getOperatorConfig, operatorConfigList } from 'config/operatorConfigs';

describe('helpers', () => {
  describe('validateCase', () => {
    it('requires arguments when exit is set', () => {
      expect(
        validateCase({
          operatorConfig: getOperatorConfig(Operators.has_any_word),
          exitName: 'My Exit',
          exitEdited: true,
        }),
      ).toMatchSnapshot();
    });

    it('does not require arguments without an exit', () => {
      expect(
        validateCase({
          operatorConfig: getOperatorConfig(Operators.has_any_word),
        }),
      ).toMatchSnapshot();
    });

    it('suggests operator exit names for no operands', () => {
      expect(
        validateCase({
          operatorConfig: getOperatorConfig(Operators.has_text),
        }),
      ).toMatchSnapshot();
    });

    it('suggests min - max exits', () => {
      expect(
        validateCase({
          operatorConfig: getOperatorConfig(Operators.has_number_between),
          min: '5',
          max: '40',
        }),
      ).toMatchSnapshot();
    });

    it('cross validates min - max', () => {
      expect(
        validateCase({
          operatorConfig: getOperatorConfig(Operators.has_number_between),
          min: '50',
          max: '3',
        }),
      ).toMatchSnapshot();
    });

    it('does not suggest an empty range for exit name, ie " - "', () => {
      expect(
        validateCase({
          operatorConfig: getOperatorConfig(Operators.has_number_between),
        }).categoryName.value,
      ).toEqual('');

      expect(
        validateCase({
          operatorConfig: getOperatorConfig(Operators.has_number_between),
        }),
      ).toMatchSnapshot();
    });
  });

  describe('prefix', () =>
    operatorConfigList.forEach(({ verboseName, type }) =>
      it(`should prefix "${verboseName}" operator appropriately`, () => {
        if (verboseName) {
          expect(prefix(type)).toMatchSnapshot();
        }
      }),
    ));

  describe('getExitName', () => {
    it('should create range exit name', () => {
      expect(
        getCategoryName({
          operatorConfig: getOperatorConfig(Operators.has_number_between),
          min: { value: '5' },
          max: { value: '10' },
        }),
      ).toBe('5 - 10');
    });

    it('should defer to edited exit names', () => {
      expect(
        getCategoryName({
          operatorConfig: getOperatorConfig(Operators.has_number_between),
          min: { value: '5' },
          max: { value: '10' },
          categoryName: { value: 'My Exit' },
          categoryNameEdited: true,
        }),
      ).toBe('My Exit');

      expect(
        getCategoryName({
          operatorConfig: getOperatorConfig(Operators.has_number),
          categoryName: { value: 'My Exit' },
          categoryNameEdited: true,
        }),
      ).toBe('My Exit');
    });

    it('should use operator names as necessary', () => {
      expect(
        getCategoryName({
          operatorConfig: getOperatorConfig(Operators.has_number),
        }),
      ).toBe('Has Number');
    });

    it('should generate names for single operands', () => {
      expect(
        getCategoryName({
          operatorConfig: getOperatorConfig(Operators.has_any_word),
          argument: { value: 'color red green blue' },
        }),
      ).toBe('Color');
    });

    it('it should generate exits for relative dates', () => {
      const testRelativeDateExit = (
        op: Operators,
        value: number,
        exitName: string,
      ): void => {
        expect(
          getCategoryName({
            operatorConfig: getOperatorConfig(op),
            argument: { value: value + '' },
          }),
        ).toBe(exitName);
      };

      testRelativeDateExit(Operators.has_date_eq, 5, 'Today + 5 days');
      testRelativeDateExit(Operators.has_date_eq, -4, 'Today - 4 days');
      testRelativeDateExit(Operators.has_date_eq, 1, 'Today + 1 day');

      testRelativeDateExit(Operators.has_date_lt, 5, 'Before today + 5 days');
      testRelativeDateExit(Operators.has_date_lt, -4, 'Before today - 4 days');
      testRelativeDateExit(Operators.has_date_lt, 1, 'Before today + 1 day');

      testRelativeDateExit(Operators.has_date_gt, 5, 'After today + 5 days');
      testRelativeDateExit(Operators.has_date_gt, -4, 'After today - 4 days');
      testRelativeDateExit(Operators.has_date_gt, 1, 'After today + 1 day');
    });

    it('should have empty categories without argument', () => {
      expect(
        getCategoryName({
          operatorConfig: getOperatorConfig(Operators.has_any_word),
        }),
      ).toBe('');
    });

    it('should have empty categories with empty argument', () => {
      expect(
        getCategoryName({
          operatorConfig: getOperatorConfig(Operators.has_any_word),
          argument: { value: '' },
        }),
      ).toBe('');
    });
  });

  describe('getMinMax', () => {
    const emptyStrs = { min: '', max: '' };

    it('should return emtpy strings if arg param is an empty array', () => {
      expect(getMinMax([])).toEqual(emptyStrs);
    });

    it("should return empty strings if arg param doesn't contain two number-containing strings", () => {
      expect(getMinMax(['a', 'b'])).toEqual(emptyStrs);
    });

    it('should return a number-containing string MIN and empty string MAX if first arg is a number-containing string and second arg is not', () => {
      expect(getMinMax(['1', 'b'])).toEqual({ min: '1', max: '' });
    });

    it('should return an empty string MIN and number-containing string MAX if first arg is not a number-containing string and second arg is', () => {
      expect(getMinMax(['a', '2'])).toEqual({ min: '', max: '2' });
    });
  });

  describe('isFloat', () => {
    it('should return true if argument is string containing a float', () => {
      ['0.2', '.2', '+.2', '+0.2', '-.2', '-0.2', '2', '2.'].forEach(arg =>
        expect(isFloat(arg as any)).toBeTruthy(),
      );
    });

    it('should return false if argument is not a string containing a float', () => {
      ['u.2', '0.2u', 'a'].forEach(arg =>
        expect(isFloat(arg as any)).toBeFalsy(),
      );
    });
  });

  describe('isInt', () => {
    it('should return true if argument is string containing an int', () => {
      ['1', '+1', '-1'].forEach(arg => expect(isInt(arg)).toBeTruthy());
    });

    it('should return false if argument is not a string containing an int', () => {
      ['0.1', 'e24', '-.3', '5+', '5-', 'a'].forEach(arg =>
        expect(isInt(arg)).toBeFalsy(),
      );
    });
  });

  describe('strContainsNum', () => {
    it('should return true if string contains only a float or int', () => {
      [
        '0.2',
        '.2',
        '+.2',
        '+0.2',
        '-.2',
        '-0.2',
        '2',
        '2.',
        '1',
        '+1',
        '-1',
      ].forEach(arg => expect(strContainsNum(arg)).toBeTruthy());
    });

    it('should return false if string does not contain only a float or int', () => {
      ['a', 'e24', '2a', '0.2u', 'u.2', 'u0.2'].forEach(arg =>
        expect(strContainsNum(arg)).toBeFalsy(),
      );
    });
  });

  describe('parseNum', () => {
    it('should return a float if passed a string containing only a float', () => {
      const float = parseNum('0.2');

      expect(typeof float === 'number' && isFinite(float)).toBeTruthy();
    });

    it('should return an int if passed a string containing only an int', () => {
      const int = parseNum('1');

      expect(typeof int === 'number' && isFinite(int)).toBeTruthy();
    });
  });
});
