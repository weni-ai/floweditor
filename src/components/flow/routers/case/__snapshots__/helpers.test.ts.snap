// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`helpers > prefix > should prefix "has a date after" operator appropriately 1`] = `"After "`;

exports[`helpers > prefix > should prefix "has a date before" operator appropriately 1`] = `"Before "`;

exports[`helpers > prefix > should prefix "has a date equal to" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has a date" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has a number above" operator appropriately 1`] = `">"`;

exports[`helpers > prefix > should prefix "has a number at or above" operator appropriately 1`] = `">= "`;

exports[`helpers > prefix > should prefix "has a number at or below" operator appropriately 1`] = `"<= "`;

exports[`helpers > prefix > should prefix "has a number below" operator appropriately 1`] = `"< "`;

exports[`helpers > prefix > should prefix "has a number between" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has a number equal to" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has a number" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has a phone number" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has a time" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has all of the words" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has an email" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has an error" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has any of the words" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has district" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has only the phrase" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has some text" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has state" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has the category" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has the phrase" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "has ward" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "is in the group" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "is not empty" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "matches regex" operator appropriately 1`] = `""`;

exports[`helpers > prefix > should prefix "starts with" operator appropriately 1`] = `""`;

exports[`helpers > validateCase > cross validates min - max 1`] = `
{
  "argument": {
    "validationFailures": [],
    "value": "",
  },
  "categoryName": {
    "validationFailures": [],
    "value": "50 - 3",
  },
  "categoryNameEdited": false,
  "confidence": {
    "validationFailures": [],
    "value": "",
  },
  "district": {
    "validationFailures": [],
    "value": "",
  },
  "intent": {
    "validationFailures": [],
    "value": null,
  },
  "max": {
    "validationFailures": [
      {
        "message": "Maximum value must be a more than the minimum",
      },
    ],
    "value": "3",
  },
  "min": {
    "validationFailures": [
      {
        "message": "Minimum value must be less than the maximum",
      },
    ],
    "value": "50",
  },
  "operatorConfig": {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
  "state": {
    "validationFailures": [],
    "value": "",
  },
  "valid": false,
}
`;

exports[`helpers > validateCase > does not require arguments without an exit 1`] = `
{
  "argument": {
    "validationFailures": [],
    "value": "",
  },
  "categoryName": {
    "validationFailures": [],
    "value": "",
  },
  "categoryNameEdited": false,
  "confidence": {
    "validationFailures": [],
    "value": "",
  },
  "district": {
    "validationFailures": [],
    "value": "",
  },
  "intent": {
    "validationFailures": [],
    "value": null,
  },
  "max": {
    "validationFailures": [],
    "value": "",
  },
  "min": {
    "validationFailures": [],
    "value": "",
  },
  "operatorConfig": {
    "operands": 1,
    "type": "has_any_word",
    "verboseName": "has any of the words",
  },
  "state": {
    "validationFailures": [],
    "value": "",
  },
  "valid": true,
}
`;

exports[`helpers > validateCase > does not suggest an empty range for exit name, ie " - " 1`] = `
{
  "argument": {
    "validationFailures": [],
    "value": "",
  },
  "categoryName": {
    "validationFailures": [],
    "value": "",
  },
  "categoryNameEdited": false,
  "confidence": {
    "validationFailures": [],
    "value": "",
  },
  "district": {
    "validationFailures": [],
    "value": "",
  },
  "intent": {
    "validationFailures": [],
    "value": null,
  },
  "max": {
    "validationFailures": [],
    "value": "",
  },
  "min": {
    "validationFailures": [],
    "value": "",
  },
  "operatorConfig": {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
  "state": {
    "validationFailures": [],
    "value": "",
  },
  "valid": true,
}
`;

exports[`helpers > validateCase > requires arguments when exit is set 1`] = `
{
  "argument": {
    "validationFailures": [
      {
        "message": "Value is required",
      },
    ],
    "value": "",
  },
  "categoryName": {
    "validationFailures": [],
    "value": "My Exit",
  },
  "categoryNameEdited": true,
  "confidence": {
    "validationFailures": [],
    "value": "",
  },
  "district": {
    "validationFailures": [],
    "value": "",
  },
  "intent": {
    "validationFailures": [],
    "value": null,
  },
  "max": {
    "validationFailures": [],
    "value": "",
  },
  "min": {
    "validationFailures": [],
    "value": "",
  },
  "operatorConfig": {
    "operands": 1,
    "type": "has_any_word",
    "verboseName": "has any of the words",
  },
  "state": {
    "validationFailures": [],
    "value": "",
  },
  "valid": false,
}
`;

exports[`helpers > validateCase > suggests min - max exits 1`] = `
{
  "argument": {
    "validationFailures": [],
    "value": "",
  },
  "categoryName": {
    "validationFailures": [],
    "value": "5 - 40",
  },
  "categoryNameEdited": false,
  "confidence": {
    "validationFailures": [],
    "value": "",
  },
  "district": {
    "validationFailures": [],
    "value": "",
  },
  "intent": {
    "validationFailures": [],
    "value": null,
  },
  "max": {
    "validationFailures": [],
    "value": "40",
  },
  "min": {
    "validationFailures": [],
    "value": "5",
  },
  "operatorConfig": {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
  "state": {
    "validationFailures": [],
    "value": "",
  },
  "valid": true,
}
`;

exports[`helpers > validateCase > suggests operator exit names for no operands 1`] = `
{
  "argument": {
    "validationFailures": [],
    "value": "",
  },
  "categoryName": {
    "validationFailures": [],
    "value": "Has Text",
  },
  "categoryNameEdited": false,
  "confidence": {
    "validationFailures": [],
    "value": "",
  },
  "district": {
    "validationFailures": [],
    "value": "",
  },
  "intent": {
    "validationFailures": [],
    "value": null,
  },
  "max": {
    "validationFailures": [],
    "value": "",
  },
  "min": {
    "validationFailures": [],
    "value": "",
  },
  "operatorConfig": {
    "categoryName": "Has Text",
    "operands": 0,
    "type": "has_text",
    "verboseName": "has some text",
  },
  "state": {
    "validationFailures": [],
    "value": "",
  },
  "valid": true,
}
`;
